[MRG+2] Add class_weight support to the forests & trees

trevorstephens opened the pull request:
This PR adds support for specifying `class_weight` in the forest classifier constructors.
- Right now it only supports single-output classification problems. It would be possible however to accept a list of dicts for each target (or just the preset string 'auto' overall), and multiply the expanded weight vectors such that a sample with two minority classes becomes even more important, etc. I couldn't find any precedent in other classifiers to guide this.
- Should an exception, warning or tantrum be thrown for an 'auto' `class_weight` used with the `warm_start` option?

Any other comments ?

trevorstephens committed:
add support for class_weights

trevorstephens committed:
add multioutput support & bootstrap auto mode

trevorstephens committed:
expanded class_weight dimension fix

trevorstephens committed:
add class_weight to trees, expand tests & minor refactor

glouppe commented:
Is this still a work in progress? I can start reviewing the PR if you feel this is more or less ready.

trevorstephens committed:
parameter validation checks & tests for errors

trevorstephens commented:
That would be great @glouppe , thanks.

Added tests for the `class_weight` parameter that raise errors when warranted. This is mostly for multi-output and valid strings as the `compute_class_weight` module does a number of checks that I am currently deferring to. I can add explicit tests for these cases if you think it is necessary. 

I warn for the string presets and `warm_start` as it appears this has slightly different usage to `SGDClassifier` from reading through the relevant discussions in the original `warm_start` issues and PR.

glouppe commented on sklearn/utils/estimator_checks.py:
```diff
@@ -737,6 +737,8 @@ def check_class_weight_classifiers(name, Classifier):
         if hasattr(classifier, "n_iter"):
             classifier.set_params(n_iter=100)
+        if hasattr(classifier, "min_weight_fraction_leaf"):
+            classifier.set_params(min_weight_fraction_leaf=0.01)
```
Why this?

trevorstephens replied:
![download](https://cloud.githubusercontent.com/assets/5210848/5615282/c1881130-94ae-11e4-89f7-40bd7a95a90f.png)

It's an extremely noisy dataset that the trees seem to have a hard time fitting as the test, I believe, is written for linear models. That was to force it to learn a very small model. I could also get it to pass by forcing a decision stump here with `max_depth=1` for instance.

amueller replied:
maybe the test could also be reworked. The test passed for rbf svms, too, I think.
I wrote some of the class_weight tests, but I feel these are very hard to do. There are some examples with carefully crafted datasets, maybe these should be used instead?

amueller replied:
Actually, this test seems to be fine to me. As long as there is any regularization, it should work. I guess the bootstrapping alone doesn't help enough as there are very few trees in the ensemble by default.

trevorstephens replied:
Yup, there's also only 2 features in this dataset too, so the randomized feature selection doesn't have much of a chance to generalize. Happy to make any mods that are deemed necessary though. Admit this is a bit of a hack to get a passing grade from Travis, though the module itself does look a bit harder through my new Iris tests in `tree` and `ensemble`.

amueller replied:
I think your fix is ok.

trevorstephens commented:
Any other comments on the PR @amueller & @glouppe ?

glouppe commented on sklearn/ensemble/forest.py:
```diff
@@ -377,8 +406,9 @@ def _set_oob_score(self, X, y):
 
-    def _validate_y(self, y):
-        y = np.copy(y)
+    def _validate_y_cw(self, y_org):
```
Can you be explicit and call the variable `y_original`?

glouppe commented:
Besides the small cosmit, I am +1 for merge. Tests are thorough. Could you also add an entry in the whatsnew file? 

Thanks for your work @trevorstephens! This is a very helpful addition :)

(and sorry for the lack of responsiveness these days...)

trevorstephens committed:
Y-org rename & whats_new update

trevorstephens commented:
Thanks @glouppe ! Requested changes are made. Second reviewer anyone (while I figure out how to get around a lovely 11th hour merge conflict) ... ?

trevorstephens committed:
Merge branch 'master' into rf-class_weight

Conflicts:
	doc/whats_new.rst

trevorstephens commented:
Travis seems to have failed for Py2.7 on an unrelated doctest for `model_selection.rst`, and failed to complete too. Can it be restarted?

amueller commented:
This is very unusual..... There really shouldn't be any travis failures any more :-/ ping @ogrisel.

amueller commented on sklearn/ensemble/forest.py:
```diff
@@ -388,7 +418,50 @@ def _validate_y(self, y):
+                                     "in class_weight should match number of "
+                                     "outputs.")
+
+            if self.class_weight != 'bootstrap' or not self.bootstrap:
```
could / should this be refactored for trees and forests?

amueller commented:
Looks good to me apart from maybe a refactoring of the class weight computation.

trevorstephens commented:
@amueller is there something you are specifically concerned about? The explicit for loop is simply over the number of outputs, which will usually be a small number, most of the time just 1. Or are you referring to `class_weight_k` being different for different options?

amueller commented:
I was not concerned about performance but code duplication.

trevorstephens commented:
Ah. I made a few comments about this a couple of weeks back: https://github.com/scikit-learn/scikit-learn/pull/3961#issuecomment-67912046 

TL;DR: It may be possible, but would result in identical calculations for every tree where this way we only do it once.

amueller commented:
never mind then, :+1: from me.

trevorstephens commented:
Great, thanks!

Any way to kick off Travis again?

GaelVaroquaux commented:
I restarted the failed job.

trevorstephens commented:
@GaelVaroquaux thanks!

trevorstephens commented:
Thanks for the reviews @amueller and @glouppe , looks like Travis is happy now, we good to merge?

GaelVaroquaux commented on sklearn/ensemble/forest.py:
```diff
@@ -211,11 +232,17 @@ def fit(self, X, y, sample_weight=None):
         self.n_outputs_ = y.shape[1]
 
-        y = self._validate_y(y)
+        y, cw = self._validate_y_cw(y)
```
Could we have a more explicite name for what "cw" means (for the variable name and the function call).

GaelVaroquaux commented on sklearn/ensemble/forest.py:
```diff
@@ -211,11 +232,17 @@ def fit(self, X, y, sample_weight=None):
 
+        if cw is not None:
+            if sample_weight is not None:
+                sample_weight *= cw
```
Don't we risk modifying an input argument here?

trevorstephens committed:
rename vars & copy sample_weight

trevorstephens commented:
@GaelVaroquaux , thanks for looking it over! Let me know what you think of the latest.

trevorstephens commented:
@amueller @glouppe @GaelVaroquaux & others

I've been thinking about rolling `class_weight` out to the other ensembles in a future PR (so as not to undo the much-appreciated reviews so far) and have an API question. Forests use a bootstrap sample, but `GradientBoostingClassifier` uses sampling without replacement, while the `BaggingClassifier` allows both bootstrapping and sampling without replacement (`AdaBoostClassifier` doesn't implement any of these, so no issue there). Should the current `class_weight='bootstrap'` option for `RandomForestClassifier` be renamed to something less specific for consistency across ensemble classifiers? Maybe `'subsample'`, `'sample'`, `'estimator'` or something else?

Sorry to bring this up after flipping from [WIP], but it seems important to decide on before merge given the other estimators in the module.

GaelVaroquaux commented:
@trevorstephens : that's a very good comment. Maybe I would use 'subsample', which I find is less confusing than 'estimator'.

GaelVaroquaux commented on sklearn/ensemble/forest.py:
```diff
@@ -211,11 +234,17 @@ def fit(self, X, y, sample_weight=None):
 
+        if expanded_class_weight is not None:
+            if sample_weight is not None:
+                sample_weight = np.copy(sample_weight) * expanded_class_weight
```
The 'np.copy' doesn't seem necessary above.

GaelVaroquaux commented on sklearn/ensemble/forest.py:
```diff
@@ -89,6 +87,29 @@ def _parallel_build_trees(tree, forest, X, y, sample_weight, tree_idx, n_trees,
+                weight_k = np.array([weight_k[np.where(classes_boot == c)][0]
+                                     if c in classes_boot
+                                     else 0.
+                                     for c in classes_full])
```
A clever use of np.choose should enable removing this for loop and making the corresponding code much faster (hint: use mode='clip').

trevorstephens replied:
@GaelVaroquaux are you referring to the `for k in range(y.shape[1]):` loop or the list comprehension? I can make the weight_k calc a lot prettier with:

    # Get class weights for the bootstrap sample
    weight_k = np.choose(classes_full,
                         compute_class_weight('auto', np.unique(y_boot), y_boot), 
                         mode='clip')

in place of L98-104 certainly makes the code nicer. Is this close to what you were aiming for?

GaelVaroquaux replied:
> @GaelVaroquaux are you referring to the for k in range(y.shape[1]): loop or the
> list comprehension? I can make the weight_k calc a lot prettier with:

> # Get class weights for the bootstrap sample
> weight_k = np.choose(classes_full,
>                      compute_class_weight('auto', np.unique(y_boot), y_boot),
>                      mode='clip')

> in place of L98-104 certainly makes the code nicer. Is this close to what you
> were aiming for?

Yes. It should also be significantly faster.

trevorstephens replied:
I had to also use `np.searchsorted` here in order for it to be properly interpreted for all cases, but it looks a lot nicer than the list comprehension. Thanks for the tip!

trevorstephens committed:
Merge branch 'master' into rf-class_weight

Conflicts:
	doc/whats_new.rst
	sklearn/ensemble/forest.py
	sklearn/tree/tree.py

trevorstephens committed:
rename cw option to subsample & refactor its implementation

trevorstephens commented:
Thanks for the comments @GaelVaroquaux . Renamed the `class_weight='bootstrap'` option to `'subsample'` and implemented the other changes you suggested. Let me know what you think.

GaelVaroquaux commented:
LGTM.

Two :+1: and my review. I think that this can be merged.

Merging. Thanks!

trevorstephens commented:
Thanks! Cheers!
